src/stateManager/utils/FraudProofService.ts

```diff
@@ -0,0 +1,178 @@
+    ) {
+        super("Invalid leader fraud detected");
+    }
+}
+
+// ────────────────────── FRAUD PROOF SERVICE ─────────────────────
+
+/**
+ * Service class for handling fraud proof creation and validation
+ */
+export class FraudProofService {
+    constructor(private readonly storage: Storage) {}
+
+    /**
+     * Create invalid state transition proof
+     */
+    createInvalidStateTransitionProof(
+        previousEntity: PreviousEntity,
+        signedBlock: SignedBlockStruct
+    ): BlockInvalidStateTransitionProofStruct {
+        let prevSignedBlock: SignedBlockStruct | undefined;
+        let prevStateSnapshot: StateSnapshot;
+
+        if (previousEntity.blockConfirmation) {
+            // Height > 0 case - we have a previous block
+            prevSignedBlock = previousEntity.blockConfirmation.signedBlock;
+            const prevBlock = Block.decode(prevSignedBlock!.encodedBlock);
+            prevStateSnapshot =
+                this.storage.stateSnapshots.getStateSnapshotByHash(
+                    prevBlock.stateSnapshotHash
+                )!;
+        } else {
+            // Height === 0 case - we have genesis state snapshot
+            prevSignedBlock = NULL as unknown as SignedBlockStruct;
+            prevStateSnapshot = previousEntity.stateSnapshot!;
+        }
+
+        return {
+            invalidBlock: signedBlock,
+            previousBlock: prevSignedBlock,
+            previousBlockStateSnapshot: prevStateSnapshot.toStruct(),
+            previousStateStateMachineState:
+                this.storage.stateMachineStates.getStateMachineState(
+                    prevStateSnapshot.stateMachineStateHash
+                )!
+        };
+    }
+
+    /**
+     * Create invalid timestamp proof
+     */
+    createInvalidTimestampProof(
+        previousEntity: PreviousEntity,
+        signedBlock: SignedBlockStruct
+    ): InvalidTimestampProofStruct {
+        let prevSignedBlock: SignedBlockStruct | undefined;
+        let prevStateSnapshot: StateSnapshot;
+
+        if (previousEntity.blockConfirmation) {
+            // Height > 0 case - we have a previous block
+            prevSignedBlock = previousEntity.blockConfirmation.signedBlock;
+            const prevBlock = Block.decode(prevSignedBlock!.encodedBlock);
+            prevStateSnapshot =
+                this.storage.stateSnapshots.getStateSnapshotByHash(
+                    prevBlock.stateSnapshotHash
+                )!;
+        } else {
+            // Height === 0 case - we have genesis state snapshot
+            prevSignedBlock = NULL as unknown as SignedBlockStruct;
+            prevStateSnapshot = previousEntity.stateSnapshot!;
+        }
+
+        return {
+            invalidBlock: signedBlock,
+            previousBlock: prevSignedBlock,
+            previousStateSnapshot: prevStateSnapshot.toStruct()
+        };
+    }
+
+    /**
+     * Fraud proof creators mapped by exception name
+     */
+    private readonly fraudProofCreators = {
```

Comment:
when taken into account with my previous comment above this entire section with fraud proof errors is over engineering